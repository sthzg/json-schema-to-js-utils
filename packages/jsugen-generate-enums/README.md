# `jsugen-generate-enums`

## Usage

```
const jsugenGenerateEnums = require('jsugen-generate-enums');

// TODO: DEMONSTRATE API
```
