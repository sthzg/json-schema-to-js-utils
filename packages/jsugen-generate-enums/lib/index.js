import { filter, map, reduce, tap } from 'rxjs/operators';
import {
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
  EMPTY_STRING,
  byMemberDefinitionIsEnum,
  toTemplateRawStringReducer,
  withCompileToTemplate,
  withPrependToString,
  withPrettier,
  withWrite,
  enrichWithPathNodeVars,
} from '@sthzg/jsugen-core';
import { fromJsonSchema } from '@sthzg/jsugen-core/lib/sources/jsonSchema';
import { buildTemplateVars } from './buildTemplateVars';
import { template as enumModuleTemplate } from './enum.tpl';

export function generate({ schema, out }) {
  // ---
  // Configure Transformer Factories.
  // ---
  const compileToTemplate = withCompileToTemplate(enumModuleTemplate);
  const prependHeaders = withPrependToString(DEFAULT_FILE_DOCSTRING);
  const prettify = withPrettier(DEFAULT_PRETTIER_OPTIONS);
  const write = withWrite(out);

  // ---
  // Observable.
  // ---
  return fromJsonSchema(schema).pipe(
    /* Filtering */
    filter(byMemberDefinitionIsEnum),

    /* Templating */
    map(enrichWithPathNodeVars),
    map(buildTemplateVars),
    map(compileToTemplate),

    /* To String */
    reduce(toTemplateRawStringReducer, EMPTY_STRING),
    map(prependHeaders),
    map(prettify),

    /* Output */
    tap(write),
  );
}
