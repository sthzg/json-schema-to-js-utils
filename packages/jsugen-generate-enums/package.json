{
  "name": "@sthzg/jsugen-generate-enums",
  "version": "0.0.1",
  "description": "> TODO: description",
  "author": "Stephan Herzog <sthzgvie@gmail.com>",
  "homepage": "https://github.com/sthzg/jsugen#readme",
  "license": "MIT",
  "main": "lib/index.js",
  "directories": {
    "lib": "lib"
  },
  "files": [
    "lib"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sthzg/jsugen.git"
  },
  "bugs": {
    "url": "https://github.com/sthzg/jsugen/issues"
  },
  "dependencies": {
    "@sthzg/jsugen-core": "^0.0.1",
    "lodash.camelcase": "^4.3.0",
    "lodash.upperfirst": "^4.3.1"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  }
}
