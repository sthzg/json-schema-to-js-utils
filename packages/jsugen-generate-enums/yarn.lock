# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


lodash.camelcase@^4.3.0:
  version "4.3.0"
  resolved "https://registry.yarnpkg.com/lodash.camelcase/-/lodash.camelcase-4.3.0.tgz#b28aa6288a2b9fc651035c7711f65ab6190331a6"
  integrity sha1-soqmKIorn8ZRA1x3EfZathkDMaY=

lodash.upperfirst@^4.3.1:
  version "4.3.1"
  resolved "https://registry.yarnpkg.com/lodash.upperfirst/-/lodash.upperfirst-4.3.1.tgz#1365edf431480481ef0d1c68957a5ed99d49f7ce"
  integrity sha1-E2Xt9DFIBIHvDRxolXpe2Z1J984=
