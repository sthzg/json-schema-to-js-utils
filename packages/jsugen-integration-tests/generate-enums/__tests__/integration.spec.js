import { generate } from '@sthzg/jsugen-generate-enums';
import schema from '../../resources/example.schema';
import contactSchema from '../../resources/example.contact';

describe('generate-enums', () => {
  test('should pass full integration test', done => {
    expect.assertions(1);

    generate({ schema }).subscribe(output => {
      expect(output).toMatchSnapshot();
      done();
    });
  });

  test('should pass example.contact.json integration test', done => {
    expect.assertions(1);

    generate({ schema: contactSchema }).subscribe(output => {
      expect(output).toMatchSnapshot();
      done();
    });
  });
});
