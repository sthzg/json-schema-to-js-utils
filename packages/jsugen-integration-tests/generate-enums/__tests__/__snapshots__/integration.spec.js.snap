// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate-enums should pass example.contact.json integration test 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const GenderValues = ['female', 'male', 'unspecified'];
export const Gender = {
  FEMALE: 'female',
  MALE: 'male',
  UNSPECIFIED: 'unspecified',
};

export const PhoneTypeValues = [
  'PRIVATE_LANDLINE',
  'PRIVATE_MOBILE',
  'WORK_LANDLINE',
  'WORK_MOBILE',
];
export const PhoneType = {
  PRIVATE_LANDLINE: 'PRIVATE_LANDLINE',
  PRIVATE_MOBILE: 'PRIVATE_MOBILE',
  WORK_LANDLINE: 'WORK_LANDLINE',
  WORK_MOBILE: 'WORK_MOBILE',
};
"
`;

exports[`generate-enums should pass full integration test 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const Level1Member4Values = [
  'A_SHOULD_BE_SORTED_AS_FIRST',
  'Z_SHOULD_BE_SORTED_AS_LAST',
];
export const Level1Member4 = {
  A_SHOULD_BE_SORTED_AS_FIRST: 'A_SHOULD_BE_SORTED_AS_FIRST',
  Z_SHOULD_BE_SORTED_AS_LAST: 'Z_SHOULD_BE_SORTED_AS_LAST',
};

export const Level1Member8Level2Member2Values = ['ONE.BAM', 'ONE.BAZ'];
export const Level1Member8Level2Member2 = {
  'ONE.BAM': 'ONE.BAM',
  'ONE.BAZ': 'ONE.BAZ',
};

export const Level1Member9Values = ['BAR', 'FOO'];
export const Level1Member9 = {
  BAR: 'BAR',
  FOO: 'FOO',
};
"
`;
